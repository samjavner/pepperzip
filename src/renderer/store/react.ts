import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
