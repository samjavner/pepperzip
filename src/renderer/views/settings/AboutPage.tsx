import * as React from "react";

export const AboutPage: React.FunctionComponent = () => <div>About</div>;

export default AboutPage;
