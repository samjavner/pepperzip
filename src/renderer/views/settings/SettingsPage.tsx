import * as React from "react";

export const SettingsPage: React.FunctionComponent = () => <div>Settings</div>;

export default SettingsPage;
