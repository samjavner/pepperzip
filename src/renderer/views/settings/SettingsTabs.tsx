import { faCog, faInfoCircle } from "@fortawesome/free-solid-svg-icons";
import * as React from "react";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";
import { Navtab, Navtabs } from "../../components/Navtabs";
import { GlobalState } from "../../core/model";
import * as navigation from "../../core/navigation";

interface SettingsTabsProps {
    activePage: navigation.SettingsPage;
    onSettingsClick: () => void;
    onAboutClick: () => void;
}

export const SettingsTabs: React.SFC<SettingsTabsProps> = props => (
    <Navtabs>
        <Navtab
            isActive={props.activePage === "settings"}
            icon={faCog}
            text="Settings"
            onClick={props.onSettingsClick}
        />
        <Navtab
            isActive={props.activePage === "about"}
            icon={faInfoCircle}
            text="About"
            onClick={props.onAboutClick}
        />
    </Navtabs>
);

const mapStateToProps = (state: GlobalState) => ({
    activePage: navigation.selectors.getActiveSettingsPage(state),
});

const mapDispatchToProps = (dispatch: Dispatch) =>
    bindActionCreators(
        {
            onSettingsClick: () =>
                navigation.actions.setActiveSettingsPage("settings"),
            onAboutClick: () =>
                navigation.actions.setActiveSettingsPage("about"),
        },
        dispatch
    );

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(SettingsTabs);
