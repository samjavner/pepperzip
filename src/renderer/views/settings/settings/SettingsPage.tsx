import * as React from "react";
import { connect } from "react-redux";
import { GlobalState } from "../../../core/model";

export const SettingsPage: React.SFC = props => <div>Settings</div>;

const mapStateToProps = (state: GlobalState) => ({});

export default connect(mapStateToProps)(SettingsPage);
