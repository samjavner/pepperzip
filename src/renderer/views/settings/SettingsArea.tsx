import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import * as actions from "../../app/actions";
import { State } from "../../app/state";
import AboutPage from "./about/AboutPage";
import SettingsPage from "./settings/SettingsPage";
import SettingsTabs from "./SettingsTabs";

function mapStateToProps(state: State) {
    return {
        activePage: state.activeSettingsPage,
    };
}

function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {};
}

type SettingsAreaProps = ReturnType<typeof mapStateToProps> &
    ReturnType<typeof mapDispatchToProps>;

const SettingsArea: React.SFC<SettingsAreaProps> = props => (
    <>
        <SettingsTabs />
        {props.activePage === "settings" && <SettingsPage />}
        {props.activePage === "about" && <AboutPage />}
    </>
);

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(SettingsArea);
