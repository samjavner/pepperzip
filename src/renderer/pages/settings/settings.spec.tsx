import * as React from "react";
import * as renderer from "react-test-renderer";
import * as settings from "./settings";

describe("settings", () => {
    describe("Page", () => {
        it("renders the text 'Settings'", () => {
            const component = renderer.create(<settings.Page />);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});
