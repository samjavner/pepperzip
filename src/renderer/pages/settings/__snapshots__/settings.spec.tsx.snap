// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings Page renders the text 'Settings' 1`] = `
<div>
  Settings
</div>
`;
